polynote-kernel/src/main/scala/polynote/kernel/interpreter/scal/ScalaCompleter.scala

```diff
@@ -222,26 +185,125 @@ class ScalaCompleter[Compiler <: ScalaCompiler](
+    val candidate = stats.filterNot(tree => tree.pos == null || !tree.pos.isDefined || tree.pos.isTransparent || !tree.pos.properlyIncludes(pos))
+      .headOption
+
+    candidate.flatMap {
+      baseTree => baseTree.collect {
+        case tree if tree.pos != null && tree.pos.isDefined && tree.pos.end == pos.point =>
+          val a = tree.attachments.get[OriginalPos]
+          tree
+      }.lastOption
+    }
+  }
+
+  // the compiler's completionsAt method has a bug; it causes an exception if the name is empty
+  // (e.g. `foo.`) which is a pretty common case. So there will be a fair amount of copypasta here,
+  // as the component methods we'd need are all private. This bit is copied from interactive.Global#typeMembers
+  private def typeMembers(context: Context, tree: Tree) = {
+
+    // had to copypasta this class as well, since it's private.
+    class Members[M <: Member] extends mutable.LinkedHashMap[Name, Set[M]] {
+      import scala.reflect.internal.Flags._
+      override def default(key: Name) = Set()
+
+      private def matching(sym: Symbol, symtpe: Type, ms: Set[M]): Option[M] = ms.find { m =>
+        (m.sym.name == sym.name) && (m.sym.isType || (m.tpe matches symtpe))
+      }
+
+      private def keepSecond(m: M, sym: Symbol, implicitlyAdded: Boolean): Boolean =
+        m.sym.hasFlag(ACCESSOR | PARAMACCESSOR) &&
+          !sym.hasFlag(ACCESSOR | PARAMACCESSOR) &&
+          (!implicitlyAdded || m.implicitlyAdded)
+
+      def add(sym: Symbol, pre: Type, implicitlyAdded: Boolean)(toMember: (Symbol, Type) => M) {
+        if ((sym.isGetter || sym.isSetter) && sym.accessed != NoSymbol) {
+          add(sym.accessed, pre, implicitlyAdded)(toMember)
+        } else if (!sym.name.decodedName.containsName("$") && !sym.isError && !sym.isArtifact && sym.hasRawInfo) {
+          val symtpe = pre.memberType(sym) onTypeError ErrorType
+          matching(sym, symtpe, this(sym.name)) match {
+            case Some(m) =>
+              if (keepSecond(m, sym, implicitlyAdded)) {
+                //print(" -+ "+sym.name)
+                this(sym.name) = this(sym.name) - m + toMember(sym, symtpe)
+              }
+            case None =>
+              //print(" + "+sym.name)
+              this(sym.name) = this(sym.name) + toMember(sym, symtpe)
+          }
         }
-        depth += 1
-        super.traverse(tree)
-        depth -= 1
       }
+
+      def addNonShadowed(other: Members[M]): Unit = {
+        for ((name, ms) <- other)
+          if (ms.nonEmpty && this(name).isEmpty) this(name) = ms
+      }
+
+      def allMembers: List[M] = values.toList.flatten
     }
-    traverser.traverse(topTree)
-    deepest match {
-      case ValDef(_, _, _, rhs) => rhs
-      case tree => tree
+
+    val superAccess = tree.isInstanceOf[Super]
+    val members = new Members[TypeMember]
+    def addTypeMember(sym: Symbol, pre: Type, inherited: Boolean, viaView: Symbol): Unit = {
```

Comment:
Most of this junk was copypasta from the compiler's private methods. Their usage of them has a blocker in it (will comment if I find the right spot)